File: frontend/src/lib/api/manufacturing.ts
/**
 * Manufacturing API Client
 * Handles all API calls for BOM and Production
 */

import api from '@/lib/api';

export interface BOM {
  id: number;
  product_id: number;
  product_name?: string;
  version: number;
  status: 'draft' | 'active' | 'archived';
  effective_date?: string;
  expiry_date?: string;
  notes?: string;
  created_by?: string;
  created_at: string;
  lines?: BOMLine[];
}

export interface BOMLine {
  id: number;
  bom_id: number;
  component_id: number;
  component_name?: string;
  component_code?: string;
  quantity: number;
  unit: string;
  waste_percent: number;
  sequence: number;
}

export interface ProductionOrder {
  id: number;
  bom_id: number;
  bom_name?: string;
  quantity: number;
  status: 'planned' | 'in_progress' | 'completed' | 'cancelled';
  cost_center_id?: number;
  start_date?: string;
  end_date?: string;
  actual_start_date?: string;
  actual_end_date?: string;
  actual_hours?: number;
  labor_rate?: number;
  notes?: string;
  created_by?: string;
  created_at: string;
}

export interface CreateBOMData {
  product_id: number;
  version: number;
  effective_date?: string;
  notes?: string;
  lines: {
    component_id: number;
    quantity: number;
    unit: string;
    waste_percent?: number;
    sequence?: number;
  }[];
}

export interface CreateProductionOrderData {
  bom_id: number;
  quantity: number;
  cost_center_id?: number;
  start_date?: string;
  end_date?: string;
  labor_rate?: number;
  notes?: string;
}

export async function fetchBOMs(status?: string): Promise<BOM[]> {
  const params = status ? { status } : {};
  const response = await api.get('/api/manufacturing/bom', { params });
  return response.data;
}

export async function fetchBOM(id: number): Promise<BOM> {
  const response = await api.get(`/api/manufacturing/bom/${id}`);
  return response.data;
}

export async function createBOM(data: CreateBOMData): Promise<BOM> {
  const response = await api.post('/api/manufacturing/bom', data);
  return response.data;
}

export async function activateBOM(id: number): Promise<BOM> {
  const response = await api.post(`/api/manufacturing/bom/${id}/activate`);
  return response.data;
}

export async function fetchProductionOrders(status?: string): Promise<ProductionOrder[]> {
  const params = status ? { status } : {};
  const response = await api.get('/api/manufacturing/orders', { params });
  return response.data;
}

export async function createProductionOrder(data: CreateProductionOrderData): Promise<ProductionOrder> {
  const response = await api.post('/api/manufacturing/orders', data);
  return response.data;
}

File: backend/app/models/manufacturing.py
"""Manufacturing models - BOM and Production"""
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, DECIMAL, Text, Date
from sqlalchemy.orm import relationship
from datetime import datetime, timezone

from app.models.base import Base


class BOMHeader(Base):
    """Bill of Materials header"""
    __tablename__ = "bom_headers"

    id = Column(Integer, primary_key=True, index=True)
    company_id = Column(Integer, ForeignKey("companies.id", ondelete="CASCADE"), nullable=False, index=True)
    product_id = Column(Integer, nullable=False, index=True)
    version = Column(Integer, nullable=False)
    status = Column(String(20), default="draft")
    effective_date = Column(Date, nullable=True)
    expiry_date = Column(Date, nullable=True)
    notes = Column(Text, nullable=True)
    created_by = Column(String(255), nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    lines = relationship("BOMLine", back_populates="bom", cascade="all, delete-orphan")


class BOMLine(Base):
    """BOM component line"""
    __tablename__ = "bom_lines"

    id = Column(Integer, primary_key=True, index=True)
    bom_id = Column(Integer, ForeignKey("bom_headers.id", ondelete="CASCADE"), nullable=False, index=True)
    component_id = Column(Integer, nullable=False)
    quantity = Column(DECIMAL(15, 3), nullable=False)
    unit = Column(String(20), nullable=False)
    waste_percent = Column(DECIMAL(5, 2), default=0)
    sequence = Column(Integer, default=0)

    bom = relationship("BOMHeader", back_populates="lines")


class ProductionOrder(Base):
    """Production job order"""
    __tablename__ = "production_orders"

    id = Column(Integer, primary_key=True, index=True)
    company_id = Column(Integer, ForeignKey("companies.id", ondelete="CASCADE"), nullable=False, index=True)
    bom_id = Column(Integer, ForeignKey("bom_headers.id"), nullable=False, index=True)
    quantity = Column(DECIMAL(15, 3), nullable=False)
    status = Column(String(20), default="planned")
    cost_center_id = Column(Integer, nullable=True)
    start_date = Column(Date, nullable=True)
    end_date = Column(Date, nullable=True)
    actual_start_date = Column(Date, nullable=True)
    actual_end_date = Column(Date, nullable=True)
    actual_hours = Column(DECIMAL(10, 2), nullable=True)
    labor_rate = Column(DECIMAL(10, 2), nullable=True)
    notes = Column(Text, nullable=True)
    created_by = Column(String(255), nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))


class ProductionMaterial(Base):
    """Material issued to production"""
    __tablename__ = "production_materials"

    id = Column(Integer, primary_key=True, index=True)
    production_order_id = Column(Integer, ForeignKey("production_orders.id", ondelete="CASCADE"), nullable=False, index=True)
    product_id = Column(Integer, nullable=False)
    required_qty = Column(DECIMAL(15, 3), nullable=False)
    issued_qty = Column(DECIMAL(15, 3), default=0)
    issue_date = Column(DateTime(timezone=True), nullable=True)
    issued_by = Column(String(255), nullable=True)


class ProductionOutput(Base):
    """Finished goods output"""
    __tablename__ = "production_output"

    id = Column(Integer, primary_key=True, index=True)
    production_order_id = Column(Integer, ForeignKey("production_orders.id", ondelete="CASCADE"), nullable=False, index=True)
    product_id = Column(Integer, nullable=False)
    quantity = Column(DECIMAL(15, 3), nullable=False)
    cost = Column(DECIMAL(15, 2), nullable=True)
    unit_cost = Column(DECIMAL(15, 2), nullable=True)
    output_date = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    recorded_by = Column(String(255), nullable=True)


class ScrapRecord(Base):
    """Production scrap/waste"""
    __tablename__ = "scrap_records"

    id = Column(Integer, primary_key=True, index=True)
    production_order_id = Column(Integer, ForeignKey("production_orders.id", ondelete="CASCADE"), nullable=False, index=True)
    product_id = Column(Integer, nullable=False)
    quantity = Column(DECIMAL(15, 3), nullable=False)
    reason = Column(String(100), nullable=True)
    recorded_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    recorded_by = Column(String(255), nullable=True)
